stats/statsvariables.cpp

```diff
@@ -0,0 +1,1674 @@
+TemperatureBinner air_temperature_binner10(10, true);
+TemperatureBinner air_temperature_binner20(20, true);
+
+struct TemperatureType : public StatsVariableTemplate<StatsVariable::Type::Numeric> {
+	// Generate 2, 5, 10 and 20-units binners in the constructor
+	TemperatureBinner b2, b5, b10, b20;
+	TemperatureType(bool air) : b2(2, air), b5(5, air), b10(10, air), b20(20, air)
+	{
+	}
+	QString unitSymbol() const override {
+		return get_temp_unit();
+	}
+	int decimals() const override {
+		return 1;
+	}
+	double tempToFloat(temperature_t t) const {
+		if (t.mkelvin <= 0)
+			return invalid_value<double>();
+		return prefs.units.temperature == units::CELSIUS ?
+				mkelvin_to_C(t.mkelvin) : mkelvin_to_F(t.mkelvin);
+	}
+	std::vector<const StatsBinner *> binners() const override {
+		return { &b2, &b5, &b10, &b20 };
+	}
+	std::vector<StatsOperation> supportedOperations() const override {
+		return { StatsOperation::Median, StatsOperation::Mean, StatsOperation::TimeWeightedMean };
+	}
+};
+
+struct WaterTemperatureType : TemperatureType {
+	WaterTemperatureType() : TemperatureType(false)
+	{
+	}
+	QString name() const override {
+		return StatsTranslations::tr("Water temperature");
+	}
+	double toFloat(const dive *d) const override {
+		return tempToFloat(d->watertemp);
+	}
+};
+
+struct AirTemperatureType : TemperatureType {
+	AirTemperatureType() : TemperatureType(true)
+	{
+	}
+	QString name() const override {
+		return StatsTranslations::tr("Air temperature");
+	}
+	double toFloat(const dive *d) const override {
+		return tempToFloat(d->airtemp);
+	}
+};
+
+// ============ Weight, binned in 1, 2, 5, 10 m or 2, 4, 10 or 20 ft bins ============
+
+struct WeightBinner : public IntRangeBinner<WeightBinner, IntBin> {
+	bool metric;
+	WeightBinner(int bin_size, bool metric) : IntRangeBinner(bin_size), metric(metric)
+	{
+	}
+	QString name() const override {
+		QLocale loc;
+		return StatsTranslations::tr("in %1 %2 steps").arg(loc.toString(bin_size),
+								   get_weight_unit());
+	}
+	QString unitSymbol() const override {
+		return get_weight_unit();
+	}
+	int to_bin_value(const dive *d) const {
+		return metric ? total_weight(d) / 1000 / bin_size
+			      : lrint(grams_to_lbs(total_weight(d))) / bin_size;
+	}
+};
+
+static WeightBinner weight_binner_1kg(1, true);
+static WeightBinner weight_binner_2kg(2, true);
+static WeightBinner weight_binner_5kg(5, true);
+static WeightBinner weight_binner_10kg(10, true);
+static WeightBinner weight_binner_2lbs(2, false);
+static WeightBinner weight_binner_5lbs(4, false);
+static WeightBinner weight_binner_10lbs(10, false);
+static WeightBinner weight_binner_20lbs(20, false);
+
+struct WeightType : public StatsVariableTemplate<StatsVariable::Type::Numeric> {
+	QString name() const override {
+		return StatsTranslations::tr("Weight");
+	}
+	QString unitSymbol() const override {
+		return get_weight_unit();
+	}
+	int decimals() const override {
+		return prefs.units.weight == units::KG ? 1 : 0;
```

Comment:
I always thought this was an odd choice (and it's been around forever)

Reply:
I think it kind of makes sense, because standard leads come in half-pound and full-kg units, no?

Reply:
not really. leads almost never weigh what people claim they do... and they are typically multiples of US pounds or half KG. So 10 lbs are 4.5kg. And because Linus and I both 'thought' metric a lot, but dove mostly with US weights, we wanted it to look correct both ways. Which is of course dumb, because if you dive with 3kg in Egypt then that's 6.6lbs :-)

Reply:
Oh, I just realize that it is exactly the other way around than I would have expected. Should this condition be inverted?

Reply:
What you have here is what we have elsewhere. One decimal for metric, no decimal for pounds. It's inconsistent and stupid. We should lead the way and simply always offer one decimal :-)

Reply:
I inverted the condition: it is now one decimal for kg and no decimals for lbs. Of course we might simply change to one decimal fraction for all.

Reply:
> What you have here is what we have elsewhere. One decimal for metric, no decimal for pounds. 

Which is funny, because I'm quite sure that this was just a mix-up from my side. Replaced by `return 1;`.